### ISSUE TITLE
SCANsat science points

### ISSUE THREAD

**Issue by chrbkr** [Kerbalism/Kerbalism#609]
I'm having issues with science point for SCANsat scans. First of all, there's only an enabled/disabled status in the vessel menu of Kerbalism, and it doesn't show up in Kerbalism's science menu at all.

Addditionally, there's something off in the R&D archives as well. For example, I did a complete multispectral scan of Kerbin, Mun and Minmus. This is what I see in the archives:

![SCANsat science](https://user-images.githubusercontent.com/13020383/75371092-705af700-58c6-11ea-965e-c01b36393abf.jpg)

**Comment by valerian** [comment]
Interesting, I just realized that I have the same issues.

Could it be related to the fact that my scans got to be interrupted at some point for lack of power or storage?

**Comment by gotmachine** [comment]
Yeah the ScanSat support code that generate science is a bit hacky, that doesn't really surprise me.
We might revisit it eventually in a future release, for now that's not really a priority.

**Comment by gotmachine** [comment]
I have seen various reprots of issues with our ScanSat implementation. My own notes : 
```
this whole thing :
	// - should not be using it's own warp buffer, but use the transmit buffer drive instead
	// - should not disable itself when there is no ec, but stay enabled and scale output by ec.AvailabilityFactor
	// The current implementation will cause EC resource sim destabilization by putting it into a produce/not produce cycle
	// it works by stopping/starting the scansat module dependning on ec availability
	// Instead we should keep it enabled but reflection-scale the "covered %" by ec.AvailabilityFactor (and drive / warpdrive space)
	// Not sure how feasible this is, but I fear the changes to the resource sim will cause this to work even more unreliably than before.
	// I don't have time now to rewrite and test it, but that probably need to be done.
```
Nullref spam reported in the RO discord on ScanSat v19 by RurouniDonut :
```
[EXC 19:41:27.852] NullReferenceException: Object reference not set to an instance of an object
    KERBALISM.ScienceSituationUtils.IsAvailableForExperiment (KERBALISM.ScienceSituation situation, KERBALISM.ExperimentInfo experiment) (at <44d2b28eb09144efbf56dab50bc4321b>:0)
    KERBALISM.VesselSituations.GetExperimentSituation (KERBALISM.ExperimentInfo expInfo) (at <44d2b28eb09144efbf56dab50bc4321b>:0)
    KERBALISM.KerbalismScansat.BackgroundUpdate (Vessel vessel, ProtoPartSnapshot p, ProtoPartModuleSnapshot m, KERBALISM.KerbalismScansat kerbalismScansat, Part part_prefab, KERBALISM.VesselData vd, KERBALISM.ResourceInfo ec, System.Double elapsed_s) (at <44d2b28eb09144efbf56dab50bc4321b>:0)
    KERBALISM.Background.Update (Vessel v, KERBALISM.VesselData vd, KERBALISM.VesselResources resources, System.Double elapsed_s) (at <44d2b28eb09144efbf56dab50bc4321b>:0)
    KERBALISM.Kerbalism.FixedUpdate () (at <44d2b28eb09144efbf56dab50bc4321b>:0)
    UnityEngine.DebugLogHandler:LogException(Exception, Object)
    ModuleManager.UnityLogHandle.InterceptLogHandler:LogException(Exception, Object)
    UnityEngine.Debug:CallOverridenDebugHandler(Exception, Object)
```
Also :
https://github.com/S-C-A-N/SCANsat/issues/367

So unless we do a full rewrite of the support code, given the mountain of work we have elsewhere, my current stance on this is to remove the special scansat handling altogether. No support is better than buggy support.

**Comment by AlexanderNZ** [comment]
Hi @gotmachine - I can see you've moved this to in progress. Just wanted to check what that means? Are you in the process of fixing this or removing the code per https://github.com/Kerbalism/Kerbalism/issues/609#issuecomment-727176016 ?

The reason I ask is that I'd be keen to dive in and take a look at the code and see if there is anything I can do to improve the integration with ScanSat (i.e. move from buggy support to non-buggy support!)

I had a look on the forum thread and couldn't see if any work around this has started or stopped. What do you think?

**Comment by gotmachine** [comment]
No, I have have no plan of working on this, and I will remove the current support altogether unless I or someone else step in to rewrite it in a satisfactory way before the next release. The scansat patch is a relatively minor "extra" and isn't strictly needed for the two mods to work together.

If you want to try to fix the current 3.x code (master branch) you're free to do so. But it isn't an easy problem to fix and it involve plugging quite deep into the specifics or both Kerbalism and ScanSat.

**Comment by gotmachine** [comment]
Did some analysis of SCANsat source.

A possible way to implement proper support could be to create a Kerbalism specific module that handle EC requirements, and the is registered as a sensor by the main SCANsat scenario `SCANcontroller`.

Then, harmony-overwrite the `SCANcontroller.doScanPass()` method, so instead of directly updating the body coverage : 
https://github.com/S-C-A-N/SCANsat/blob/e2d7bde255f5a263a04f5ace60af5c674e50516a/SCANsat/SCANcontroller.cs#L2609
we save an array of [x, y, sensorType] values in a File derivative, and then update the body coverage when that file is transmitted/recovered.
